File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/Soyjakparty.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.core.site.sites;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.k1rakishou.chan.core.site.ChunkDownloaderSiteProperties;
import com.github.k1rakishou.chan.core.site.Site;
import com.github.k1rakishou.chan.core.site.SiteIcon;
import com.github.k1rakishou.chan.core.site.common.CommonSite;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanActions;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanApi;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanCommentParser;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanEndpoints;
import com.github.k1rakishou.chan.core.site.parser.CommentParserType;
import com.github.k1rakishou.common.DoNotStrip;
import com.github.k1rakishou.model.data.board.ChanBoard;
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor;
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor;

import org.jetbrains.annotations.NotNull;
import java.util.Map;

import okhttp3.HttpUrl;

@DoNotStrip
public class Soyjakparty
        extends CommonSite {
    private final ChunkDownloaderSiteProperties chunkDownloaderSiteProperties;
    public static final String SITE_NAME = "Soyjak.party";

    public static final CommonSiteUrlHandler URL_HANDLER = new CommonSiteUrlHandler() {
        private static final String ROOT = "https://soyjak.party/";

        @Override
        public Class<? extends Site> getSiteClass() {
            return Soyjakparty.class;
        }

        @Override
        public HttpUrl getUrl() {
            return HttpUrl.parse(ROOT);
        }

        @Override
        public HttpUrl[] getMediaHosts() {
            return new HttpUrl[]{getUrl()};
        }

        @Override
        public String[] getNames() {
            return new String[]{"Soyjakparty"};
        }

        @Override
        public String desktopUrl(ChanDescriptor chanDescriptor, @Nullable Long postNo) {
            if (chanDescriptor instanceof ChanDescriptor.CatalogDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .toString();
            } else if (chanDescriptor instanceof ChanDescriptor.ThreadDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .addPathSegment("res")
                        .addPathSegment(((ChanDescriptor.ThreadDescriptor) chanDescriptor).getThreadNo() + ".html")
                        .toString();
            } else {
                return null;
            }
        }
    };

    public Soyjakparty() {
        chunkDownloaderSiteProperties = new ChunkDownloaderSiteProperties(true, true);
    }

    @Override
    public void setup() {
        setEnabled(true);
        setName(SITE_NAME);
        setIcon(SiteIcon.fromFavicon(getImageLoaderV2(), HttpUrl.parse("https://soyjak.party/static/favicon.png")));

        setBoards(
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "a"), "Anime"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "b"), "Random"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "giga"), "Gigachad"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "int"), "International"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "muv"), "Music & Video Games"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "nate"), "Coal"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "pol"), "Politics"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "qa"), "Question & Answer"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "r"), "Requests"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "r9k"), "ROBOT9999"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "raid"), "Raid"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "sneed"), "Sneed"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "soy"), "Soyjak"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "suggest"), "Suggest"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "tv"), "Television & Film"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "webm"), "Videos and GIF"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "x"), "Paranormal/Schizo")
        );

        setResolvable(URL_HANDLER);

        setConfig(new CommonConfig() {
            @Override
            public boolean siteFeature(SiteFeature siteFeature) {
                return super.siteFeature(siteFeature) || siteFeature == SiteFeature.POSTING;
            }
        });

        setEndpoints(new VichanEndpoints(this, "https://soyjak.party/", "https://soyjak.party/") {
                    @Override
    public HttpUrl thumbnailUrl(
            BoardDescriptor boardDescriptor,
            boolean spoiler,
            int customSpoilter,
            Map<String, String> arg
    ) {
                String ext;
                switch (arg.get("ext")) {
                    case "jpeg":
                    case "jpg":
                    case "png":
                    case "gif":
                        ext = arg.get("ext");
                        break;
                    default:
                        ext = "jpg";
                        break;
                }

{       return root.builder()
                .s(boardDescriptor.getBoardCode())
                .s("thumb")
                .s(arg.get("tim") + "." + ext)
                .url();
    }}});
    
        setActions(new VichanActions(this, getProxiedOkHttpClient(), getSiteManager(), getReplyManager()));
        setApi(new VichanApi(getSiteManager(), getBoardManager(), this));
        setParser(new VichanCommentParser());
    }

    @NotNull
    @Override
    public CommentParserType commentParserType() {
        return CommentParserType.VichanParser;
    }

    @NonNull
    @Override
    public ChunkDownloaderSiteProperties getChunkDownloaderSiteProperties() {
        return chunkDownloaderSiteProperties;
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/vichan/soyjakparty/SoyjakParty.kt
package com.github.k1rakishou.chan.core.site.sites.vichan.soyjakparty

import com.github.k1rakishou.chan.core.site.SiteEndpoints
import com.github.k1rakishou.chan.core.site.common.vichan.VichanEndpoints
import com.github.k1rakishou.chan.core.site.sites.vichan.BaseVichanSite
import com.github.k1rakishou.model.data.board.ChanBoard
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor.CatalogDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor.ThreadDescriptor
import okhttp3.HttpUrl

class SoyjakParty : BaseVichanSite(
  defaultDomain = "https://soyjak.party"
) {
  private val boards by lazy {
    buildList {
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "q"), "the 'party"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "soy"), "soyjaks"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "jak"), "jaks"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "qa"), "question & answer"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "r"), "requests and soy art"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "caca"), "cacaborea"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "a"), "tranime"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "raid"), "raid: shadow legends"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "int"), "international"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "mtv"), "music, television, video games"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "pol"), "international politics"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "sci"), "soyence and technology"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "craft"), "minecraft"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "fnac"), "five nights at cobson's"))
      add(ChanBoard.create(BoardDescriptor.create(descriptor.siteName, "nate"), "coals"))
    }
  }

  override val enabled: Boolean = false
  override val name: String = SITE_NAME
  override val urlHandler by lazy { SoyjakPartyUrlHandler(this) }
  override val endpoints by lazy { SoyjakPartyEndpoints(this) }
  override val staticBoards = boards

  class SoyjakPartyEndpoints(
    soyjakParty: SoyjakParty
  ) : VichanEndpoints(soyjakParty) {
    override fun thumbnailUrl(
      boardDescriptor: BoardDescriptor,
      spoiler: Boolean,
      customSpoilers: Int,
      arg: Map<String, String>?
    ): HttpUrl {
      requireNotNull(arg)

      val extension = when (arg.get("ext")) {
        "jpg", "jpeg", "gif", "webp" -> "." + arg.get("ext")
        "webm", "mp4" -> ".jpg"
        else -> ".png"
      }
      return root.builder()
        .s(boardDescriptor.boardCode)
        .s("thumb")
        .s(arg.get("tim") + extension)
        .url()
    }

    override fun thread(
      threadDescriptor: ThreadDescriptor,
      contentType: SiteEndpoints.ContentType,
      archive: Boolean
    ): HttpUrl? {
      return root.builder()
        .s(threadDescriptor.boardCode())
        .s("thread")
        .s(threadDescriptor.threadNo.toString() + ".json")
        .url()
    }
  }

  class SoyjakPartyUrlHandler(soyjakParty: SoyjakParty) : CommonSiteUrlHandler(soyjakParty) {
    override fun desktopUrl(chanDescriptor: ChanDescriptor, postNo: Long?, postSubNo: Long?): String? {
      when (chanDescriptor) {
        is CatalogDescriptor -> {
          return rootUrl.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .toString()
        }
        is ThreadDescriptor -> {
          return rootUrl.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .addPathSegment("thread")
            .addPathSegment(chanDescriptor.threadNo.toString() + ".html")
            .toString()
        }
        else -> {
          return null
        }
      }
    }
  }


  companion object {
    const val SITE_NAME: String = "Soyjak.party"
  }
}
